src/main/scala/dxWDL/CompilerTopologicalSort.scala

```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        // Build the initial map of predecessors
+        val toPred = edges.foldLeft(Map[A, Set[A]]()) { (acc, e) =>
+            acc + (e._1 -> acc.getOrElse(e._1, Set())) + (e._2 -> (acc.getOrElse(e._2, Set()) + e._1))
+        }
+        tsort(toPred, Seq())
+    }
+
+    // Internal procedure to sort children nodes at a particular level in the AST
+    def tsortASTnodes(nodes: Seq[Scope], cState:State, recursionDepth: Int) : Seq[Scope] = {
+        val indent = " "*recursionDepth*4
+
+        val recursedNodes = nodes.map {
+             case scatter : Scatter => {
+                 Utils.trace(cState.verbose, indent+"Recursively performing toological sort on scatter: " + scatter.fullyQualifiedName)
+                 val newScatter = WdlRewrite.scatter(scatter,
+                    tsortASTnodes(scatter.children, cState, recursionDepth + 1))
+                 Utils.trace(cState.verbose, indent+"End recursion")
+                 newScatter.asInstanceOf[GraphNode]
+             }
+             case anyOtherNode => anyOtherNode.asInstanceOf[GraphNode]
+         }
+
+        // Create workflow graph sequence of edges for topological sorting
+        // Nodes in the graph (Scopes) can be a call, a scatter, or a declaration
+
+        // Create a mapping from a child of a scatter node to the root level scatter
+        // node it belongs to. This allows for a root level scatter node to be placed
+        // after its last dependency.
+
+        val scatterParent : Map[GraphNode, Scatter] = recursedNodes.map {
+            case scat: Scatter => scat.descendants.map { d => (d.asInstanceOf[GraphNode], scat) }
+            case _ => Set[(GraphNode, Scatter)]()
+        }.flatten.toMap
+
+        // Generate a list of edges corresponding to dependencies for this call
+        // e.g. suppose call x has dependencies a,b,c.  This procedure will generate:
+        // [(a,x), (b,x), (c,x)]
+        val edges : Set[(Scope, Scope)] = recursedNodes.map { gnode =>
+            val nodeDependencies : Vector[(Scope, Scope)] = (gnode.upstream).map { dependency =>
+               val dependencyActual : Scope =
+                   // If any node's parent is a descendant of a root
+                   // level scatter, use the scatter as the parent
+                   if (scatterParent.contains(dependency)) {
+                       scatterParent(dependency).asInstanceOf[Scope]
+                   }
+                   // Otherwise just use the actual parent
+                   else {
+                       dependency.asInstanceOf[Scope]
+                   }
+               (dependencyActual, gnode.asInstanceOf[Scope])
+            }.toVector
+
+            val descendantParents : Vector[(Scope, Scope)] = gnode match {
+                case scatter: Scatter => {
+                    // If a scatter's descendants have parents outside the scatter context,
+                    // map the scatter itself to them.
```

Comment:
You have an example for when this code is needed in TopoSort.md. Can you point to it from here?